Register OpInfos for torch.fft transforms

peterbell10 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #48712 Cleanup unnecessary SpectralFuncInfo logic
* #48428 Migrate some torch.fft tests to use OpInfos
* **#48427 Register OpInfos for torch.fft transforms**

Differential Revision: [D25266218](https://our.internmc.facebook.com/intern/diff/D25266218)

peterbell10 committed:
Register OpInfos for torch.fft transforms

[ghstack-poisoned]

peterbell10 committed:
Update on "Register OpInfos for torch.fft transforms"




[ghstack-poisoned]

peterbell10 committed:
Update on "Register OpInfos for torch.fft transforms"




[ghstack-poisoned]

peterbell10 committed:
Update on "Register OpInfos for torch.fft transforms"




[ghstack-poisoned]

peterbell10 committed:
Update on "Register OpInfos for torch.fft transforms"




[ghstack-poisoned]

peterbell10 committed:
Update on "Register OpInfos for torch.fft transforms"




[ghstack-poisoned]

peterbell10 committed:
Update on "Register OpInfos for torch.fft transforms"




[ghstack-poisoned]

peterbell10 committed:
Update on "Register OpInfos for torch.fft transforms"




[ghstack-poisoned]

peterbell10 committed:
Update on "Register OpInfos for torch.fft transforms"




[ghstack-poisoned]

mruberry commented on torch/testing/_internal/common_methods_invocations.py:
```diff
@@ -52,6 +54,19 @@ def __init__(self, input, *, args=tuple(), kwargs=None):
         self.kwargs = kwargs if kwargs is not None else {}
 
 
+_NOTHING = object()  # Unique value to distinguish default anything else
```
Add a comment describing this function and its intended use

peterbell10 replied:
Comment added.

peterbell10 committed:
Update on "Register OpInfos for torch.fft transforms"




[ghstack-poisoned]

peterbell10 committed:
Update on "Register OpInfos for torch.fft transforms"




[ghstack-poisoned]

mruberry commented on torch/testing/_internal/common_methods_invocations.py:
```diff
@@ -220,9 +237,71 @@ def sample_inputs(self, device, dtype, requires_grad=False):
+                 skips=None,
+                 **kwargs):
+        skips = skips if skips is not None else []
+        dtypesIfCPU = dtypesIfCPU if dtypesIfCPU is not None else dtypes
```
Are these necessary? Won't the OpInfo class perform this logic?

peterbell10 replied:
Fixed in #48712. As an aside, does adding small cleanup PRs onto the stack work out better than updating this PR after it's imported?

mruberry replied:
Adding the cleanup to later PRs in the stack (including creating a cleanup PR for later) is easier since editing this requires it be reimported, which will restart the internal test process.

mruberry reviewed (commented):
@peterbell10 This is awesome! I have one small code cleanup question that can be addressed in the follow-up PR. Nice work.

mruberry commented:
Sorry to bother you @peterbell10, but the internal tools are struggling with rebasing this PR. Would you try rebasing it on the latest and I'll try to get it merged ASAP? 

I tried to go through the merge conflicts internally but they're wild.

peterbell10 committed:
Update on "Register OpInfos for torch.fft transforms"


Differential Revision: [D25266218](https://our.internmc.facebook.com/intern/diff/D25266218)

[ghstack-poisoned]

peterbell10 commented:
@mruberry rebased on `viable/strict`.

peterbell10 commented:
Oh looks like some new OpInfo based tests are broken.

peterbell10 committed:
Update on "Register OpInfos for torch.fft transforms"


Differential Revision: [D25266218](https://our.internmc.facebook.com/intern/diff/D25266218)

[ghstack-poisoned]

peterbell10 committed:
Update on "Register OpInfos for torch.fft transforms"


Differential Revision: [D25266218](https://our.internmc.facebook.com/intern/diff/D25266218)

[ghstack-poisoned]

peterbell10 commented on torch/testing/_internal/jit_metaprogramming_utils.py:
```diff
@@ -229,8 +229,15 @@ def the_method({}):
+
 def get_call(method_name, func_type, args, kwargs):
-    kwargs_str = ', '.join([k + '=' + str(v) for k, v in kwargs.items()])
+    kwargs_str = ', '.join([k + '=' + value_to_literal(v) for k, v in kwargs.items()])
```
This didn't work with strings e.g. it would emit `norm=ortho` without any quotes.

mruberry replied:
Nice fix.

peterbell10 commented on torch/testing/_internal/jit_metaprogramming_utils.py:
```diff
@@ -461,12 +468,12 @@ def make_module(script):
+def check_alias_annotation(method_name, args, kwargs, *, aten_name, func_type='method'):
     formals, tensors, actuals = get_script_args(args)
-    call = get_call(method_name, 'method', actuals, kwargs)
+    call = get_call(method_name, func_type, actuals, kwargs)
```
These operators don't have a method variant so this didn't work. But since this is called in the variants loop, it should really be testing the actual.

peterbell10 commented on torch/testing/_internal/jit_metaprogramming_utils.py:
```diff
@@ -461,12 +468,12 @@ def make_module(script):
     script = script_template.format(', '.join(formals), call)
     CU = torch.jit.CompilationUnit(script)
-    torch._C._jit_check_alias_annotation(CU.the_method.graph, tuple(tensors), method_name)
+    torch._C._jit_check_alias_annotation(CU.the_method.graph, tuple(tensors), aten_name)
```
This failed because the script node is `aten::fft_fft` instead of just `"aten::" + op.name`.

peterbell10 commented on test/test_ops.py:
```diff
@@ -231,69 +231,63 @@ def test_variant_consistency_jit(self, device, dtype, op):
                     def fn(*inputs, **kwargs):
-                        attr = getattr(inputs[0], name)
-                        output = attr(*inputs[1:], **kwargs)
+                        output = func(*inputs, **kwargs)
```
Again, failed because there is no method variant. Then again, there are methods without function variants as well. Would it make sense to use `variant` here instead?

peterbell10 replied:
On second thought, `get_op()` wouldn't return anything if there was no function variant. Is `OpInfo` meant to support that case?

mruberry replied:
OpInfos weren't really designed to handle methods that don't have corresponding functions, and the few cases where this occurs in PyTorch probably just need a function variant to be added. 

This line originally requiring the method defined on torch.Tensor is probably just a historical oddity, and changing it to call the function-variant (and assuming there always is a function variant) is fine.

peterbell10 committed:
Update on "Register OpInfos for torch.fft transforms"


Differential Revision: [D25266218](https://our.internmc.facebook.com/intern/diff/D25266218)

[ghstack-poisoned]

peterbell10 committed:
Update on "Register OpInfos for torch.fft transforms"


Differential Revision: [D25266218](https://our.internmc.facebook.com/intern/diff/D25266218)

[ghstack-poisoned]

peterbell10 commented:
@mruberry PTAL.

mruberry commented on test/test_ops.py:
```diff
@@ -231,69 +231,63 @@ def test_variant_consistency_jit(self, device, dtype, op):
-                    assert variant is inplace
-                    name = op.name + "_"
-                    func_type = 'inplace'
+                name = op.name + '_' if func_type == 'inplace' else op.name
```
Not for this PR, but getting rid of these string manipulation hacks are the entire reason we're moving to OpInfos. We should probably let the name of the inplace variant of a function be requested from the OpInfo directly.

mruberry commented on test/test_ops.py:
```diff
@@ -231,69 +231,63 @@ def test_variant_consistency_jit(self, device, dtype, op):
                     #   so running it on all dtypes is would be excessive
-                    if dtype in [torch.float32, torch.int32]:
-                        check_alias_annotation(name, (*sample.input,) + sample.args, sample.kwargs)
+                    if dtype in [torch.float32, torch.int32] and func_type == 'method':
```
Why the restriction to only the method variant here?

peterbell10 replied:
Ah yes, left over from debugging so just removed.

mruberry commented on test/test_spectral_ops.py:
```diff
@@ -5,6 +5,10 @@
 import itertools
 
+# saves the torch.fft function that's clobbered by importing the torch.fft module
+fft_fn = torch.fft
```
This can be removed now that the fft deprecation PR is in.

mruberry commented on torch/testing/_internal/common_methods_invocations.py:
```diff
@@ -262,9 +282,71 @@ def sample_inputs_addmm(self, device, dtype, requires_grad):
+        if not TEST_WITH_SLOW:
+            skips.append(SkipInfo('TestGradients', 'test_fn_gradgrad'))
+
+        if not TEST_MKL:
```
These TEST_MKL and TEST_WITH_ROCM skips are cool but they can be applied universally using OpInfo's existing decorator support.

I think TEST_WITH_SLOW, however, does require special-casing like is done here, and this is probably the best way to do it.

peterbell10 replied:
Isn't this better because it saves performing any checks when running the test in the non-skipped case.

mruberry replied:
If we were worried about the performance impact you'd be correct, but I'd prefer we optimize for consistency/readability/maintainability here.

peterbell10 replied:
Okay, I replaced the `SkipInfo`s with decorators.

mruberry commented on torch/testing/_internal/common_methods_invocations.py:
```diff
@@ -262,9 +282,71 @@ def sample_inputs_addmm(self, device, dtype, requires_grad):
+                 dtypesIfCPU=None,
+                 dtypesIfCUDA=None,
+                 dtypesIfROCM=None,
+                 ndimensional,
```
Might as well add an inline comment describing that ndimensional is a bool that's true when the fft function can apply to multiple dimensions (like fftn)

mruberry commented:
@pytorchbot retest this please

Commit:
minor cleanup to retrigger CI

peterbell10 committed:
Update on "Register OpInfos for torch.fft transforms"


Differential Revision: [D25266218](https://our.internmc.facebook.com/intern/diff/D25266218)

[ghstack-poisoned]

peterbell10 committed:
Update on "Register OpInfos for torch.fft transforms"


Differential Revision: [D25266218](https://our.internmc.facebook.com/intern/diff/D25266218)

[ghstack-poisoned]

peterbell10 commented:
@mruberry latest batch of merge conflicts are resolved, PTAL.

peterbell10 committed:
Update on "Register OpInfos for torch.fft transforms"


Differential Revision: [D25266218](https://our.internmc.facebook.com/intern/diff/D25266218)

[ghstack-poisoned]

peterbell10 committed:
Update on "Register OpInfos for torch.fft transforms"


Differential Revision: [D25266218](https://our.internmc.facebook.com/intern/diff/D25266218)

[ghstack-poisoned]

peterbell10 committed:
Update on "Register OpInfos for torch.fft transforms"


Differential Revision: [D25266218](https://our.internmc.facebook.com/intern/diff/D25266218)

[ghstack-poisoned]

mruberry commented:
Finally landed! Sorry this took so long to land, @peterbell10, our internal tools really fought it.
